### ISSUE TITLE
C-Interface test returns Cuda error 200 when loading 1st kernel on Tegra X1

### ISSUE THREAD

**Issue by dammad** [NervanaSystems/nervanagpu#22]
I've compiled the C side of NervanaGPU for my Tegra X1 platform (using CUDA 7.0 in a cross-compile configuration). Yet whenever I run the C-Interface test program it gives me:

Failed to load: hgemm_nn_vec_128x128 200
Couldn't load all kernels

Error 200 appears to be CUDA_ERROR_INVALID_IMAGE. I've tried changing the architecture for NVCC from sm_50 to sm_52 but that doesn't help. 

Any ideas or is NervanaGpu incompatible with the X1 some how?

**Comment by scott-gray** [comment]
It's possible.  I've never been able to test on an X1.  Maybe compile for sm_53?

The cubin format may have changed very slightly for sm_53.  Not sure how I can debug this without an X1 to work on.  If you're ambitious, you could try exploring in maxas Cubin.pm to see what might be amiss.  nvdisasm -raw is pretty helpful as well.  The cubin is basically an elf format with very minor customizations.

**Comment by dammad** [comment]
I tried sm_53 but no joy. Any suggestions for what to look for in the nvdisasm -raw output. I'm familar with ELF but not on the Nvidia/Cuda platform. I'm happy to dump some tests for you if you like. Alas I don't have a spare X1 platform.

**Comment by scott-gray** [comment]
After a cu file is compiled with nvcc, I inject the sass code into that cubin.  There are only a handful of additional things I need to change for the gpu to accept it as a valid cubin.  I'm guessing maybe one thing may have changed.  I'd have to compare different cubins of differnt length to get some clue.

If you look in Cubin.pm you'll see a couple big comments over the getting and setting of InsCnt.  I wrote all that before I realized nvdisasm -raw more or less completely documents all the custom cubin settings.  InsCnt really should be EIATTR_EXIT_INSTR_OFFSETS.  If I was to wager, I'm betting that's the problem code.  I've already had to tweak it for cuda 7.0 so it's not surprising sm_53 also has problems.  Now I know what that data actually means so I should be able to get it right now.

But alas.. I'm pretty swamped at the moment.  It will take me a bit of time to be able to get to it.

**Comment by dammad** [comment]
To determine if the offsets are different, I compiled to a .cubin the .cu file from a simple test (the Cuda 7.0 matrixMul sample) with both sm_50 and sm_53 architectures. I used --cubin to drop the other parts. Searching through the ELF file with nvdisasm -raw shows that the sections have the same offsets. The only thing that changes are the header flags to: EF_CUDA_SM53 EF_CUDA_PTX_SM(EF_CUDA_SM53)

So alas I'm concluding the problem must be elsewhere.

**Comment by scott-gray** [comment]
That they're the same might not tell you anything.  It might just be that sm_53 is more strict about enforcing correctness.  So the fact that EIATTR_EXIT_INSTR_OFFSETS isn't being set correctly could still be the problem.  Your two examples both have the right value.

So maybe that's a good test.  Take a known working cubin and change the value of EIATTR_EXIT_INSTR_OFFSETS to see if it breaks (or you can try breaking it in other ways).  A little bit of hacking of Cubin.pm should allow you do to this.  Or you could use a hex editor.

**Comment by scott-gray** [comment]
Speaking of the X1, I'm working right now on a set of 3x3 convolution kernels that use the Winograd transform.  These should get really high utilization for minibatches down to a size of 8.  This could make the X1 a great platform for mobile inference. 

http://arxiv.org/pdf/1509.09308v1.pdf

**Comment by dammad** [comment]
Yes it might be less tolerant. I had hoped to try this today, but didn’t get time. Perhaps tomorrow.. 
On October 13, 2015 at 7:36:26 PM, Scott Gray (notifications@github.com) wrote:

That they're the same might not tell you anything. It might just be that sm_53 is more strict about enforcing correctness. So the fact that EIATTR_EXIT_INSTR_OFFSETS isn't being set correctly could still be the problem. Your two examples both have the right value.

So maybe that's a good test. Take a known working cubin and change the value of EIATTR_EXIT_INSTR_OFFSETS to see if it breaks (or you can try breaking it in other ways). A little bit of hacking of Cubin.pm should allow you do to this. Or you could use a hex editor.

**Comment by dammad** [comment]
Cool. Very useful. 

Alan.
On October 13, 2015 at 7:52:09 PM, Scott Gray (notifications@github.com) wrote:

Speaking of the X1, I'm working right now on a set of 3x3 convolution kernels that use the Winograd transform. These should get really high utilization for minibatches down to a size of 8. This could make the X1 a great platform for mobile inference.

http://arxiv.org/pdf/1509.09308v1.pdf
